client/asset/dcr/dcr.go

```diff
@@ -955,139 +979,237 @@ func (dcr *ExchangeWallet) LocktimeExpired(contract dex.Bytes) (bool, error) {
 	return time.Now().UTC().After(contractExpiry), nil
 }
 
-// FindRedemption should attempt to find the input that spends the specified
-// coin, and return the secret key if it does.
+// FindRedemption attempts to find the inputs that spend the specified coins,
+// and return the secret key for each contract when it does.
+// Every input of every block starting at the earliest mined contract block is
+// scanned in a goroutine until the spending inputs are found or an error occurs.
+// The result channel is used to notify the caller when a secret key is found for
+// a contract or if an error occurs during the search.
 //
 // NOTE: FindRedemption is necessary to deal with the case of a maker
 // redeeming but not forwarding their redemption information. The DEX does not
 // monitor for this case. While it will result in the counter-party being
 // penalized, the input still needs to be found so the swap can be completed.
-// For typical blockchains, every input of every block starting at the
-// contract block will need to be scanned until the spending input is found.
 // This could potentially be an expensive operation if performed long after
 // the swap is broadcast. Realistically, though, the taker should start
 // looking for the maker's redemption beginning at swapconf confirmations
 // regardless of whether the server sends the 'redemption' message or not.
-func (dcr *ExchangeWallet) FindRedemption(ctx context.Context, coinID dex.Bytes) (dex.Bytes, error) {
-	txHash, vout, err := decodeCoinID(coinID)
-	if err != nil {
-		return nil, err
-	}
-	txid := txHash.String()
-	// Get the wallet transaction.
-	tx, err := dcr.node.GetTransaction(txHash)
-	if err != nil {
-		if isTxNotFoundErr(err) {
-			return nil, asset.CoinNotFoundError
+func (dcr *ExchangeWallet) FindRedemption(coinIDs []dex.Bytes, resultChan chan asset.FindRedemptionResult) {
```

Comment:
I don't think this function should be asynchronous. How about accepting a `context.Context`, but letting the caller decide how to handle the long running process?

Reply:
I'd also like to move away from a manual scan and use `rpcllient.Client.Rescan` to search for spent outputs. Should be loads faster.

Reply:
Not sure how using a ctx would work for contracts that are unredeemed as of the current best block and which should wait for new blocks to scan. The latest change has the tip change monitoring goroutine performing redemption searches on new blocks.

Reply:
Yeah, wanted to get the basic setup / flow right before moving to scanning with cfilters. That would only work for dcr btw, so getting the manual scan working properly would remain beneficial for btc, while dcr can be upgraded away (likely in a follow up PR anyways) - the introduction of `dcr.findRedemptionsInBlockRange` should make the switch more straightforward.

Reply:
> Not sure how using a ctx would work for contracts that are unredeemed as of the current best block and which should wait for new blocks to scan. 

I see what you're saying. I guess the primary benefit is that it prevents unnecessary re-scanning, which is nice. A pattern that I think would simplify implementation is

1. One contract per channel.
2. Channel created and returned by the exchange wallet, along with any instant error. 

`FindRedemption(coinID dex.Bytes) (chan asset.FindRedemptionResult, error)`

Then in **core**

1. `shouldBeginFindRedemption` = `true`: Call `FindRedemption`
2. If no instant error, start a goroutine to monitor the channel
3. Mark a boolean field in `matchTracker` to indicate that monitoring is running. Check field in `shouldBeginFindRedemption`.

Reply:
That looks slick, but that would mean calling the method repeatedly for multiple contracts. Given that it's possible for multiple contracts (different matches for the same order) to be sent in the same tx, batching those contracts in one call to FindRedemption prevents fetching and scanning the same set of blocks and txs repeatedly for each contract.

So looks like there's gonna be a trade-off. Something we're okay with?

Reply:
We can try it, but the conditions under which we would gain a perf boost would be extremely rare (hopefully). My primary motivation is to keep `ExchangeWallet` implementation approachable for asset devs. Maybe batch submission is fine, but I think `FindRedemption(coinID dex.Bytes) (chan asset.FindRedemptionResult, error)` is a natural pattern lending itself to simple solutions.

Regardless, there's nothing fundamentally wrong about your approach, so I'm comfortable approving it if your really set on it. Doesn't mean I won't push for a change later though.

Reply:
Makes sense. Will try it out, see how it uncomplicates things and possibly run with it.